### ISSUE TITLE
Pen pressure function and Thank you very much

### ISSUE THREAD

**Issue by TangXiaoLin-CN** [moonlight-stream/moonlight-qt#1247]
Moonlight-Sunshine 10-point multi-touch feature is really great. I never thought I'd be using a high-performance Windows desktop with a touch gesture on a handheld device. I'm using a Surface pro 7 10-point touch tablet computer and it works very well. As of today, I have uninstalled all the GFE that I have used before because of this overwhelming function

Additionally... Is it possible to support the pressure sensing pen function? If this is supported, I believe it will be an innovative program in the field of graphics works. (especially 3D graphics)

Anyway, this 5.0 version update is another life-changing opportunity for me. Thank you so much.

**Comment by Andres-3DArtist** [comment]
Yeah please, I'm 3D artist and I'm looking forward to be able to use my main workspace remotely with Zbrush or Substance painter. Being able to use a Wacom from the client pc or Apple/Samsung/Xiaomi pencil would be awesome.

That would be a life changing

**Comment by TangXiaoLin-CN** [comment]
> Yeah please, I'm 3D artist and I'm looking forward to be able to use my main workspace remotely with Zbrush or Substance painter. Being able to use a Wacom from the client pc or Apple/Samsung/Xiaomi pencil would be awesome.
> 
> That would be a life changing

I downloaded and view code of moonlight, I find it support pen pressure function. but it is not compiled to latest release. And I find moonlight cant do more, because SDL2 just surpport pen pressure.SDL2 cant surpport pen float on screen surface as a mouse. It mean pen cant normal works on windows,except moonlight to improve it or upgrade to SDL3.

**Comment by Andres-3DArtist** [comment]
And do you think it'd be possible to update to SLD3? I mean, if we can have hope hahaha

**Comment by TangXiaoLin-CN** [comment]
> And do you think it'd be possible to update to SLD3? I mean, if we can have hope hahaha

hahah,I personally hope it is possible very much.But it is diffcult after I know status of SDL3. SDL3 is review now, and it has a lot of changed,therefore upgrade project may is complex and need much time. 

reletively,to individually changed SDL2's code for adapt windows' pen pressure is easy, but this is not conducive to upgrade SDL3.

however, official version of SDL3 will be publish in furture, project upgrade also is inevitable. We just need wait and hope.

**Comment by cgutman** [comment]
Yes, the problem is that SDL doesn't expose pen pressure until SDL 3 (which isn't yet released). Someone posted a WIP PR that converts Moonlight to SDL 3 (https://github.com/moonlight-stream/moonlight-qt/pull/1182) but it's probably outdated now since SDL3 is still making breaking changes.

Once SDL 3 is officially released and Moonlight has been updated to support it, the actual changes to support pen pressure are trivial.

**Comment by TangXiaoLin-CN** [comment]
> Yes, the problem is that SDL doesn't expose pen pressure until SDL 3 (which isn't yet released). Someone posted a WIP PR that converts Moonlight to SDL 3 (https://github.com/moonlight-stream/moonlight-qt/pull/1182) but it's probably outdated now since SDL3 is still making breaking changes.
> 
> Once SDL 3 is officially released and Moonlight has been updated to support it, the actual changes to support pen pressure are trivial.

Thank you for your reply and  hard work.

**Comment by gschintgen** [comment]
This issue has me wondering: Is a digital pen / stylus currently supposed to work correctly with absolute coordinates? (Pressure sensitivity would be a luxury at this point...)

Currently my Wacom stylus (integrated into a Thinkpad Yoga S1) is working perfectly fine locally, but if I use that Thinkpad as Moonlight client to my Sunshine host, the host only receives *relative* mouse movement packets from the pen. I.e. if I lift the pen so that it's no longer detected as hovering and then bring it near the screen again that new absolute position is not transmitted to Sunshine.

Is that expected?

(That is with moonlight 5.0.1, but since the release notes for 6.0 don't mention anything pen related I wouldn't expect changes. Host and client are Ubuntu 22.04.)

**Comment by Lucalme** [comment]
@gschintgen 

> This issue has me wondering: Is a digital pen / stylus currently supposed to work correctly with absolute coordinates? (Pressure sensitivity would be a luxury at this point...)
> 
> Currently my Wacom stylus (integrated into a Thinkpad Yoga S1) is working perfectly fine locally, but if I use that Thinkpad as Moonlight client to my Sunshine host, the host only receives _relative_ mouse movement packets from the pen. I.e. if I lift the pen so that it's no longer detected as hovering and then bring it near the screen again that new absolute position is not transmitted to Sunshine.
> 
> Is that expected?
> 
> (That is with moonlight 5.0.1, but since the release notes for 6.0 don't mention anything pen related I wouldn't expect changes. Host and client are Ubuntu 22.04.)

I'm on windows 11 on both client and host and have the exact same issue. Did you find a way to make it work?

**Comment by gschintgen** [comment]
> 
> I'm on windows 11 on both client and host and have the exact same issue. Did you find a way to make it work?

No, it's not currently supported on PC as it seems: https://github.com/LizardByte/Sunshine/pull/2606#issuecomment-2172107978

I can confirm though that Samsung's S-Pen on a tablet is working indeed, but I couldn't test it in-depth to check whether or not pressure sensitivity is working. (Caveat: the pen stutters heavily every 20-40 seconds if used via moonlight/sunshine; there is an issue about it that may have been closed prematurely. I'll have to properly test all of this sometime.)

**Comment by Andres-3DArtist** [comment]
> > I'm on windows 11 on both client and host and have the exact same issue. Did you find a way to make it work?
> 
> No, it's not currently supported on PC as it seems: [LizardByte/Sunshine#2606 (comment)](https://github.com/LizardByte/Sunshine/pull/2606#issuecomment-2172107978)
> 
> I can confirm though that Samsung's S-Pen on a tablet (moonlight-android) is working indeed (host: Sunshine Linux), but I couldn't test it in-depth to check whether or not pressure sensitivity is working. (Caveat: the pen stutters heavily every 20-40 seconds if used via moonlight/sunshine; there is an issue about it that may have been closed prematurely. I'll have to properly test all of this sometime.)

I've tested it in my Samsung S9 Ultra and it's weird. I do have pen pressure in Photoshop but not in Zbrush

**Comment by Lucalme** [comment]
> > I'm on windows 11 on both client and host and have the exact same issue. Did you find a way to make it work?
> 
> No, it's not currently supported on PC as it seems: [LizardByte/Sunshine#2606 (comment)](https://github.com/LizardByte/Sunshine/pull/2606#issuecomment-2172107978)
> 
> I can confirm though that Samsung's S-Pen on a tablet (moonlight-android) is working indeed (host: Sunshine Linux), but I couldn't test it in-depth to check whether or not pressure sensitivity is working. (Caveat: the pen stutters heavily every 20-40 seconds if used via moonlight/sunshine; there is an issue about it that may have been closed prematurely. I'll have to properly test all of this sometime.)

I'm wondering if it would be possible to write a transparent overlay with third party library (preferably linux and windows compatible) that would capture pen data and then transmit it to the app to send over network. Would have to look into code but seems it could be good temporary fix since SDL 3 is not yet stable.

**Comment by Kuuh4** [comment]
Update! SDL 3 Is out! 
https://github.com/libsdl-org/SDL/milestone/10

**Comment by Fuchsiae** [comment]
It might interest some as a temporary solution, you can use Weylus (and in particular the community fork) : https://github.com/electronstudio/WeylusCommunityEdition

Then you can play around like I did, 

1. In Sunshine deactivate all the options about touch input or pen support, restart Sunshine
2. launch the Weylus server on the host 
3. Open Moonlight on the client device
4. open your web browser above it > open the Weylus client page (<host-IP>:port) > on the side menu, uncheck the mouse and touch transmission but keep Stylus transmission checked
5. Still on the side menu, cut the video stream, and put it in black screen, power save mode for the least delay. 
(You could use only Weylus, but I find that the video stream delay is way too big compared to Moonlight, which is why I use both at the same time instead, but Weylus limited to registering pen information).
6.  Then open a software that allows you to render your current window completely transparent (I used a software called "Spooky View" on Windows) 
7. Be sure to have Moonlight open on focus, and the web browser on the Weylus client page, transparent, above it.
8. Make the browser full screen

> Et voilà ! You can use your pen directly on the screen of the host computer through your tablet device, with pen pressure registered and few delay.

**Comment by guerillaengineer** [comment]
update on this? Looks like SDL3 is out for a bit now. I'm going back to college for an associate's in CAD/CAM and design and would love to be able to use my pens, as a physically disabled person. I will start tinkering with the pr that was noted above, in the meantime.

**Comment by rbaszak** [comment]
I'm interested in this issue as well. My use case is:

Moonlight Streaming client installed on Macbook. Used with iPad as second display via Sidecar, so this way I can use Apple Pencil and pen pressure on MacOS client, but currently it works as a mouse.

**Comment by trungnguyenarts** [comment]
<img width="1480" height="979" alt="Image" src="https://github.com/user-attachments/assets/176cf5cf-b68c-4275-a9a8-e829b79a0d26" />
works fine on my ipad, the pen recognizes the pressure sensitivity however it seems to flicker the mouse cursor when it detects the pen icon and cursor at the same time.

**Comment by Tetracell** [comment]
For everyone mentioning how it works on their Android / iPad tablets, please make note that this is talking about the Windows version of Moonlight.

**Comment by Thecreatre** [comment]
anyword here?

**Comment by trungnguyenarts** [comment]
> anyword here?
It seems this is the case even for developers Moonlight and Artemis, who focus solely on games. Therefore, I developed my own fork version, focusing on 3D artist/2D artist software, compatible with all desktop Windows/Mac platforms.
https://www.reddit.com/r/MoonlightStreaming/comments/1sablql/comment/oedf4bs/

**Comment by KaizenCris** [comment]
Hi, someone with more experience on the project. Any guidelines on how support for pen tablets could be implemented?
I'm primarily a software dev, but I also draw and sculpt a bit. So being able to use blender on a more powerful pc while sculpting on my Android tablet or just a cheap laptop sounds really useful and powerful for any artist. (I mainly use a XPPen Magic Drawing Pad, but I also own a Cintiq to test with)

We could even implement multi-finger gestures! (zoom in, zoom out, rotate, quick access buttons) That would be amazing! I would like to dive more into this, but I'm new so I don't know where to start.
